pkg/api/validation/validation.go

```diff
@@ -555,15 +555,80 @@ func validateEnv(vars []api.EnvVar) errs.ValidationErrorList {
 		vErrs := errs.ValidationErrorList{}
 		if len(ev.Name) == 0 {
 			vErrs = append(vErrs, errs.NewFieldRequired("name"))
-		}
-		if !util.IsCIdentifier(ev.Name) {
+		} else if !util.IsCIdentifier(ev.Name) {
 			vErrs = append(vErrs, errs.NewFieldInvalid("name", ev.Name, cIdentifierErrorMsg))
 		}
+		vErrs = append(vErrs, validateEnvVarSources(ev)...)
 		allErrs = append(allErrs, vErrs.PrefixIndex(i)...)
 	}
 	return allErrs
 }
 
+func validateEnvVarSources(ev api.EnvVar) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	if ev.EnvVarSource == nil {
+		return allErrs
+	}
+
+	numSources := 0
+
+	switch {
+	case ev.ObjectInfo != nil:
+		numSources++
+		allErrs = append(allErrs, validateObjectInfo(ev.ObjectInfo)...)
+	case ev.HostInfo != nil:
+		numSources++
+		allErrs = append(allErrs, validateHostInfo(ev.HostInfo)...)
+	}
+
+	if ev.Value != "" && numSources >= 1 {
+		allErrs = append(allErrs, errs.NewFieldInvalid("", "", "sources cannot be specified when value is set"))
+	} else if numSources > 1 {
+		allErrs = append(allErrs, errs.NewFieldInvalid("", "", "at most one source can be specified when value is empty"))
+	}
+
+	return allErrs
+}
+
+func validateObjectInfo(ref *api.ObjectReference) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	if ref.FieldPath == "" {
+		allErrs = append(allErrs, errs.NewFieldRequired("fieldPath"))
+	}
+	if ref.APIVersion == "" {
+		allErrs = append(allErrs, errs.NewFieldRequired("apiVersion"))
+	}
+	if ref.Kind != "" {
```

Comment:
wouldn't we do better to just ignore extra fields as if the struct did not have them?